src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void StartRequest(RequestCompletionSource state)
```

Comment:
I need some clarifications on Option 2. libcurl is going to spin up a new thread for every write callback. And the data cannot be buffrered so we have to wait for a reader. @stephentoub, is the blocking you are referring to the one in StartRequest and the task that was spawned by it?